### ISSUE TITLE
Logstash hangs on Non-OK response code from Elasticsearch - logs are lost!

### ISSUE THREAD

**Issue by jordansissel** [logstash-plugins/logstash-output-elasticsearch#136]
(This issue was originally filed by @kosfar at https://github.com/elastic/logstash/issues/2509)

---

Hi,
I am using logstash 1.4.2 and elasticsearch output with protocol => "http". Our ES cluster is version 0.90.13 and we use redis as a queue mechanism. 
logstash --> redis --> logstash --> ES 

In random times we experience logstash indexer freezes, redis queue length increases and the software logs:
{:timestamp=>"2015-02-03T18:01:07.257000+0200", :message=>"Failed to flush outgoing items", :outgoing_count=>71, :exception=>#<RuntimeError: Non-OK response code from Elasticsearch: 404>, :backtrace=>["/opt/logstash/lib/logstash/outputs/elasticsearch/protocol.rb:127:in `bulk_ftw'", "/opt/logstash/lib/logstash/outputs/elasticsearch/protocol.rb:80:in`bulk'", "/opt/logstash/lib/logstash/outputs/elasticsearch.rb:315:in `flush'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/stud-0.0.17/lib/stud/buffer.rb:219:in`buffer_flush'", "org/jruby/RubyHash.java:1339:in `each'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/stud-0.0.17/lib/stud/buffer.rb:216:in`buffer_flush'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/stud-0.0.17/lib/stud/buffer.rb:193:in `buffer_flush'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/stud-0.0.17/lib/stud/buffer.rb:112:in`buffer_initialize'", "org/jruby/RubyKernel.java:1521:in `loop'", "/opt/logstash/vendor/bundle/jruby/1.9/gems/stud-0.0.17/lib/stud/buffer.rb:110:in`buffer_initialize'"], :level=>:warn}

Logstash comes back in life after process restart but the problem is that logs that are waiting in redis seem to land to nowhere, they are gone. 
In the last time, problems started at 18:01:07, stopped at 19:56:10 after logstash restart but logs that managet to get indexed are only after 19:34 (verified this on kibana). The logs between 18:01:07 and 19:34 were lost. You can check this on the screenshot. I have noticed that there is already a related open issue here:
https://logstash.jira.com/browse/LOGSTASH-2203

Can you elaborate on this plz?

![screenshot from 2015-02-03 20 09 52](https://cloud.githubusercontent.com/assets/701533/6041822/8256d2e8-ac88-11e4-9f58-94d01c3251d0.png)

**Comment by GustavoHoyer** [comment]
What field are you using for @timestamp? Maybe the @timestamp is the time when the event was processed by logstash, not the original timestamp of your source logs.

**Comment by andrewvc** [comment]
Apologies for the delay. This has been fixed in multiple ways since this was reported. Closing.

**Comment by uday3110** [comment]
@andrewvc  what multiple ways was this fixed in ? I have the same issue and would like a solution

**Comment by andrewvc** [comment]
@uday3110 this plugin has basically been rewritten since then, including all connection handling and retry logic.

If something is still going wrong we'd love to hear about it. Please open a new issue with whatever error log messages you're getting.
